facade/src/main/java/org/jboss/pnc/facade/providers/BuildConfigurationProviderImpl.java

```diff
@@ -51,4 +160,165 @@ public BuildConfigurationProviderImpl(BuildConfigurationRepository repository,
+
+        ValidationBuilder
+                .validateObject(WhenCreatingNew.class)
+                .validateAgainstRepository(repository, buildConfigurationId, true);
+
+        org.jboss.pnc.model.BuildConfiguration buildConfiguration = repository.queryById(buildConfigurationId);
+
+        org.jboss.pnc.model.BuildConfiguration clonedBuildConfiguration = buildConfiguration.clone();
+        clonedBuildConfiguration = repository.save(clonedBuildConfiguration);
+
+        logger.debug("Cloned saved BuildConfiguration: {}", clonedBuildConfiguration);
+
+        return mapper.toDTO(clonedBuildConfiguration);
+    }
+
+    @Override
+    public void addDependency(Integer configId, Integer dependencyId) {
+
+        org.jboss.pnc.model.BuildConfiguration buildConfig = repository.queryById(configId);
+        org.jboss.pnc.model.BuildConfiguration dependency = repository.queryById(dependencyId);
+
+        ValidationBuilder.validateObject(buildConfig, WhenUpdating.class)
+                .validateCondition(buildConfig != null, "No build config exists with id: " + configId)
+                .validateCondition(dependency != null, "No dependency build config exists with id: " + dependencyId)
+                .validateCondition(!configId.equals(dependencyId), "A build configuration cannot depend on itself")
+                .validateCondition(!dependency.getAllDependencies().contains(buildConfig), "Cannot add dependency from : "
+                        + configId + " to: " + dependencyId + " because it would introduce a cyclic dependency");
+
+        System.out.println("didn't throw any validation errors");
+        buildConfig.addDependency(dependency);
+        repository.save(buildConfig);
+    }
+
+    @Override
+    public void removeDependency(Integer configId, Integer dependencyId) {
+
+        org.jboss.pnc.model.BuildConfiguration buildConfig = repository.queryById(configId);
+        org.jboss.pnc.model.BuildConfiguration dependency = repository.queryById(dependencyId);
+
+        ValidationBuilder.validateObject(buildConfig, WhenUpdating.class)
+                .validateCondition(buildConfig != null, "No build config exists with id: " + configId)
+                .validateCondition(dependency != null, "No dependency build config exists with id: " + dependencyId);
+
+        buildConfig.removeDependency(dependency);
+        repository.save(buildConfig);
+    }
+
+    @Override
+    public void setProductVersion(Integer configId, Integer productVersionId) {
+
+        org.jboss.pnc.model.BuildConfiguration buildConfig = repository.queryById(configId);
+
+        ValidationBuilder.validateObject(buildConfig, WhenUpdating.class)
+                .validateCondition(buildConfig!=null, "No build config exists with id: " + configId);
+
+        ProductVersion productVersion = null;
+
+        if (productVersionId != null) {
+            productVersion = productVersionRepository.queryById(productVersionId);
```

Comment:
What happens if the product version with such ID doesn't exists? Shouldn't there be a validation?

Reply:
Will add!